engines/terraform/resolve.go

```diff
@@ -84,13 +57,111 @@ func (td *TerraformDeployment) resolveValue(intentName string, value interface{}
+	switch specRef.Source {
+	case "infra":
+		if len(specRef.Path) < 2 {
+			return nil, fmt.Errorf("infra token requires at least 2 path components")
+		}
+
+		refName := specRef.Path[0]
+		attribute := strings.Join(specRef.Path[1:], ".")
+
+		infraResource, err := td.resolveInfraResource(refName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve infrastructure resource %s: %w", refName, err)
+		}
+
+		result := infraResource.Get(jsii.String(attribute))
+		if result == nil {
+			return nil, fmt.Errorf("attribute '%s' not found on infrastructure resource '%s' (type: %s)", attribute, refName, *infraResource.Node().Id())
+		}
+		return result, nil
+
+	case "self":
+		if len(specRef.Path) < 1 {
+			return nil, fmt.Errorf("self token requires at least 1 path component")
+		}
+
+		varName := specRef.Path[0]
+
+		tfVariable, ok := td.instancedTerraformVariables[intentName][varName]
+		if !ok {
+			return nil, fmt.Errorf("variable %s does not exist for provided blueprint", varName)
+		}
+
+		return tfVariable.Value(), nil
+
+	case "var":
+		if len(specRef.Path) < 1 {
+			return nil, fmt.Errorf("var token requires at least 1 path component")
+		}
+
+		varName := specRef.Path[0]
+
+		tfVariable, ok := td.getPlatformVariable(varName)
+		if !ok {
+			return nil, fmt.Errorf("variable %s does not exist for this platform", varName)
+		}
+
+		return tfVariable.Value(), nil
+
+	default:
+		return nil, fmt.Errorf("unknown token source: %s", specRef.Source)
+	}
+}
+
+func (td *TerraformDeployment) resolveTokenString(intentName string, input string) (interface{}, error) {
+	tokens := findAllTokens(input)
+
+	if len(tokens) == 0 {
+		return input, nil
+	}
+
+	if len(tokens) == 1 && isOnlyToken(input) {
+		specRef, err := SpecReferenceFromToken(tokens[0].Token)
+		if err != nil {
+			return input, nil
+		}
+
+		return td.resolveToken(intentName, specRef)
+	}
+
+	result := input
+	for i := len(tokens) - 1; i >= 0; i-- {
+		token := tokens[i]
+
+		specRef, err := SpecReferenceFromToken(token.Token)
+		if err != nil {
+			continue
+		}
+
+		tokenValue, err := td.resolveToken(intentName, specRef)
+		if err != nil {
+			return nil, err
+		}
+
+		// Convert token to string representation for concatenation
+		replacement := cdktf.Token_AsString(tokenValue, nil)
+		if replacement == nil {
+			return nil, fmt.Errorf("failed to convert token %s to string", token.Token)
+		}
+
+		result = result[:token.Start] + *replacement + result[token.End:]
+	}
+
+	return result, nil
+}
+
 func (td *TerraformDeployment) resolveTokensForModule(intentName string, resource *ResourceBlueprint, module cdktf.TerraformHclModule) error {
 	for property, value := range resource.Properties {
 		resolvedValue, err := td.resolveValue(intentName, value)
 		if err != nil {
 			return fmt.Errorf("failed to resolve property %s for %s: %w", property, intentName, err)
 		}
-		module.Set(jsii.String(property), resolvedValue)
+		if resolvedValue != nil {
```

Comment:
This is the nil check that fixes:
https://linear.app/nitric/issue/NIT-339/build-panic